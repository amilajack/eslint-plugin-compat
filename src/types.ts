import { APIKind } from "ast-metadata-inferer/lib/types";
import { Rule } from "eslint";
import { TargetNameMappings } from "./constants";

export type BrowserListConfig =
  | string
  | Array<string>
  | {
      production?: Array<string>;
      development?: Array<string>;
    }
  | null;

// @TODO Replace with types from ast-metadata-inferer
// Types from ast-metadata-inferer
type AstMetadataApi = {
  type?: string;
  name?: string;
  object: string;
  astNodeType: "MemberExpression" | "CallExpression" | "NewExpression";
  property?: string;
  protoChainId: string;
  protoChain: Array<string>;
};

export interface Target {
  target: keyof TargetNameMappings;
  parsedVersion: number;
  version: number | string | "all";
}

export type HandleFailingRule = (
  node: AstMetadataApiWithTargetsResolver,
  eslintNode: ESLintNode
) => void;

export type TargetNames = Array<string>;

export type ESLintNode = {
  name: string;
  type: string;
  value?: unknown;
  object?: ESLintNode;
  parent?: ESLintNode;
  expression?: ESLintNode;
  property?: ESLintNode;
  callee?: ESLintNode & {
    name: string;
    type?: string;
  };
};

export interface AstMetadataApiWithTargetsResolver extends AstMetadataApi {
  id: string;
  caniuseId?: string;
  kind?: APIKind;
  getUnsupportedTargets: (
    node: AstMetadataApiWithTargetsResolver,
    targets: Target[]
  ) => Array<string>;
}

export interface Context extends Rule.RuleContext {
  settings: {
    targets?: string[];
    browsers?: Array<string>;
    polyfills?: Array<string>;
    lintAllEsApis?: boolean;
  };
}
