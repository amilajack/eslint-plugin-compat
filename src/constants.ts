export interface TargetNameMappings {
  chrome: "Chrome";
  firefox: "Firefox";
  safari: "Safari";
  ios_saf: "iOS Safari";
  ie: "IE";
  ie_mob: "IE Mobile";
  edge: "Edge";
  baidu: "Baidu";
  electron: "Electron";
  blackberry_browser: "Blackberry Browser";
  edge_mobile: "Edge Mobile";
  and_uc: "Android UC Browser";
  and_chrome: "Android Chrome";
  and_firefox: "Android Firefox";
  and_webview: "Android Webview";
  and_samsung: "Samsung Browser";
  and_opera: "Opera Android";
  opera: "Opera";
  opera_mini: "Opera Mini";
  opera_mobile: "Opera Mobile";
  node: "Node.js";
  kaios: "KaiOS";
}

// Maps an ID to the full name user will see
// E.g. during error, user will see full name instead of ID
export const STANDARD_TARGET_NAME_MAPPING: Readonly<TargetNameMappings> = {
  chrome: "Chrome",
  firefox: "Firefox",
  safari: "Safari",
  ios_saf: "iOS Safari",
  ie: "IE",
  ie_mob: "IE Mobile",
  edge: "Edge",
  baidu: "Baidu",
  electron: "Electron",
  blackberry_browser: "Blackberry Browser",
  edge_mobile: "Edge Mobile",
  and_uc: "Android UC Browser",
  and_chrome: "Android Chrome",
  and_firefox: "Android Firefox",
  and_webview: "Android Webview",
  and_samsung: "Samsung Browser",
  and_opera: "Opera Android",
  opera: "Opera",
  opera_mini: "Opera Mini",
  opera_mobile: "Opera Mobile",
  node: "Node.js",
  kaios: "KaiOS",
};

export enum AstNodeTypes {
  MemberExpression = "MemberExpression",
  CallExpression = "CallExpression",
  NewExpression = "NewExpression",
}
