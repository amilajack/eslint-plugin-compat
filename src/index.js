"use strict";
/**
 * Step 1) Entry point of plugin. Exports itself for eslint to use
 * @author Amila Welihinda
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.rules = exports.config = exports.configs = void 0;
//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------
const recommended_1 = __importDefault(require("./config/recommended"));
//------------------------------------------------------------------------------
// Plugin Definition
//------------------------------------------------------------------------------
// import all rules in lib/rules
const compat_1 = __importDefault(require("./rules/compat"));
exports.configs = {
    recommended: recommended_1.default,
};
// Kept for backwards compatibility
exports.config = exports.configs;
exports.rules = {
    compat: compat_1.default,
};
