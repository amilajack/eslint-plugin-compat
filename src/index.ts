/**
 * Step 1) Entry point of plugin. Exports itself for eslint to use
 * @author Amila Welihinda
 */

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------
import recommended from "./config/recommended";
import pkg from "../package.json";
import type { Linter } from "eslint";

//------------------------------------------------------------------------------
// Plugin Definition
//------------------------------------------------------------------------------

// import all rules in lib/rules
import compat from "./rules/compat";

const rules = {
  compat,
};

const plugin = {
  meta: {
    name: pkg.name,
    version: pkg.version,
  },
  configs: {},
  rules,
};

const configs = {
  "flat/recommended": {
    plugins: { compat: plugin },
    ...recommended.flat,
  } as Linter.FlatConfig,
  recommended: {
    plugins: ["compat"],
    ...recommended.legacy,
  } as Linter.Config,
};
plugin.configs = configs;

export = {
  ...(plugin as typeof plugin & { configs: typeof configs }),
  // TODO: Remove this in next major release
  /**
   * @deprecated Use `.configs` instead. This will be removed in the next major release.
   */
  config: configs,
};
