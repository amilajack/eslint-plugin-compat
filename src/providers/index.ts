/*
 * Step 3) Compat use CanIUse and MDN providers to check if a target browser supports a particular API
 */
import canIUseNodes from "./caniuse-provider";
import mdnNodes from "./mdn-provider";
import type { AstMetadataApiWithTargetsResolver } from "../types";

export const nodes: Array<AstMetadataApiWithTargetsResolver> = [
  ...canIUseNodes,
  ...mdnNodes,
];
