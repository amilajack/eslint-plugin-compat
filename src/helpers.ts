/* eslint no-nested-ternary: off */
import browserslist from "browserslist";
import {
  AstMetadataApiWithTargetsResolver,
  ESLintNode,
  BrowserListConfig,
  Target,
  HandleFailingRule,
  Context,
} from "./types";
import { TargetNameMappings } from "./constants";

/*
3) Figures out which browsers user is targeting

- Uses browserslist config and/or targets defined eslint config to discover this
- For every API ecnountered during traversal, gets compat record for that
- Protochain (e.g. 'document.querySelector')
  - All of the rules have compatibility info attached to them
- Each API is given to versioning.ts with compatibility info
*/
function isInsideIfStatement(context: Context) {
  return context.getAncestors().some((ancestor) => {
    return ancestor.type === "IfStatement";
  });
}

function checkNotInsideIfStatementAndReport(
  context: Context,
  handleFailingRule: HandleFailingRule,
  failingRule: AstMetadataApiWithTargetsResolver,
  node: ESLintNode
) {
  if (!isInsideIfStatement(context)) {
    handleFailingRule(failingRule, node);
  }
}

export function lintCallExpression(
  context: Context,
  handleFailingRule: HandleFailingRule,
  rules: AstMetadataApiWithTargetsResolver[],
  node: ESLintNode
) {
  if (!node.callee) return;
  const calleeName = node.callee.name;
  const failingRule = rules.find((rule) => rule.object === calleeName);
  if (failingRule)
    checkNotInsideIfStatementAndReport(
      context,
      handleFailingRule,
      failingRule,
      node
    );
}

export function lintNewExpression(
  context: Context,
  handleFailingRule: HandleFailingRule,
  rules: Array<AstMetadataApiWithTargetsResolver>,
  node: ESLintNode
) {
  if (!node.callee) return;
  const calleeName = node.callee.name;
  const failingRule = rules.find((rule) => rule.object === calleeName);
  if (failingRule)
    checkNotInsideIfStatementAndReport(
      context,
      handleFailingRule,
      failingRule,
      node
    );
}

export function lintExpressionStatement(
  context: Context,
  handleFailingRule: HandleFailingRule,
  rules: AstMetadataApiWithTargetsResolver[],
  node: ESLintNode
) {
  if (!node?.expression?.name) return;
  const failingRule = rules.find(
    (rule) => rule.object === node?.expression?.name
  );
  if (failingRule)
    checkNotInsideIfStatementAndReport(
      context,
      handleFailingRule,
      failingRule,
      node
    );
}

function protoChainFromMemberExpression(node: ESLintNode): string[] {
  if (!node.object) return [node.name];
  const protoChain = (() => {
    switch (node.object.type) {
      case "NewExpression":
      case "CallExpression":
        return protoChainFromMemberExpression(node.object.callee!);
      default:
        return protoChainFromMemberExpression(node.object);
    }
  })();
  return [...protoChain, node.property!.name];
}

export function lintMemberExpression(
  context: Context,
  handleFailingRule: HandleFailingRule,
  rules: Array<AstMetadataApiWithTargetsResolver>,
  node: ESLintNode
) {
  if (!node.object || !node.property) return;
  if (
    !node.object.name ||
    node.object.name === "window" ||
    node.object.name === "globalThis"
  ) {
    const rawProtoChain = protoChainFromMemberExpression(node);
    const [firstObj] = rawProtoChain;
    const protoChain =
      firstObj === "window" || firstObj === "globalThis"
        ? rawProtoChain.slice(1)
        : rawProtoChain;
    const protoChainId = protoChain.join(".");
    const failingRule = rules.find(
      (rule) => rule.protoChainId === protoChainId
    );
    if (failingRule) {
      checkNotInsideIfStatementAndReport(
        context,
        handleFailingRule,
        failingRule,
        node
      );
    }
  } else {
    const objectName = node.object.name;
    const propertyName = node.property.name;
    const failingRule = rules.find(
      (rule) =>
        rule.object === objectName &&
        (rule.property == null || rule.property === propertyName)
    );
    if (failingRule)
      checkNotInsideIfStatementAndReport(
        context,
        handleFailingRule,
        failingRule,
        node
      );
  }
}

export function reverseTargetMappings<K extends string, V extends string>(
  targetMappings: Record<K, V>
): Record<V, K> {
  const reversedEntries = Object.entries(targetMappings).map((entry) =>
    entry.reverse()
  );
  return Object.fromEntries(reversedEntries);
}

/**
 * Determine the targets based on the browserslist config object
 * Get the targets from the eslint config and merge them with targets in browserslist config
 * Eslint target config will be deprecated in 4.0.0
 *
 * @param configPath - The file or a directory path to look for the browserslist config file
 */
export function determineTargetsFromConfig(
  configPath: string,
  config?: BrowserListConfig
): Array<string> {
  const browserslistOpts = { path: configPath };

  const eslintTargets = (() => {
    // Get targets from eslint settings
    if (Array.isArray(config) || typeof config === "string") {
      return browserslist(config, browserslistOpts);
    }
    if (config && typeof config === "object") {
      return browserslist(
        [...(config.production || []), ...(config.development || [])],
        browserslistOpts
      );
    }
    return [];
  })();

  if (browserslist.findConfig(configPath)) {
    // If targets are defined in ESLint and browerslist configs, merge the targets together
    if (eslintTargets.length) {
      const browserslistTargets = browserslist(undefined, browserslistOpts);
      return Array.from(new Set(eslintTargets.concat(browserslistTargets)));
    }
  } else if (eslintTargets.length) {
    return eslintTargets;
  }

  // Get targets fron browserslist configs
  return browserslist(undefined, browserslistOpts);
}

/**
 * Parses the versions that are given by browserslist. They're
 *
 * ```ts
 * parseBrowsersListVersion(['chrome 50'])
 *
 * {
 *   target: 'chrome',
 *   parsedVersion: 50,
 *   version: '50'
 * }
 * ```
 * @param targetslist - List of targest from browserslist api
 * @returns - The lowest version version of each target
 */
export function parseBrowsersListVersion(
  targetslist: Array<string>
): Array<Target> {
  return (
    // Sort the targets by target name and then version number in ascending order
    targetslist
      .map((e: string): Target => {
        const [target, version] = e.split(" ") as [
          keyof TargetNameMappings,
          number | string
        ];

        const parsedVersion: number = (() => {
          if (typeof version === "number") return version;
          if (version === "all") return 0;
          return version.includes("-")
            ? parseFloat(version.split("-")[0])
            : parseFloat(version);
        })();

        return {
          target,
          version,
          parsedVersion,
        };
      }) // Sort the targets by target name and then version number in descending order
      // ex. [a@3, b@3, a@1] => [a@3, a@1, b@3]
      .sort((a: Target, b: Target): number => {
        if (b.target === a.target) {
          // If any version === 'all', return 0. The only version of op_mini is 'all'
          // Otherwise, compare the versions
          return typeof b.parsedVersion === "string" ||
            typeof a.parsedVersion === "string"
            ? 0
            : b.parsedVersion - a.parsedVersion;
        }
        return b.target > a.target ? 1 : -1;
      }) // First last target always has the latest version
      .filter(
        (e: Target, i: number, items: Array<Target>): boolean =>
          // Check if the current target is the last of its kind.
          // If it is, then it's the most recent version.
          i + 1 === items.length || e.target !== items[i + 1].target
      )
  );
}
