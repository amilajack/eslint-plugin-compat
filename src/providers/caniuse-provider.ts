import * as lite from "caniuse-lite";
import { STANDARD_TARGET_NAME_MAPPING, AstNodeTypes } from "../constants";
import { AstMetadataApiWithTargetsResolver, Target } from "../types";

/**
 * Take a target's id and return it's full name by using `targetNameMappings`
 * ex. {target: and_ff, version: 40} => 'Android FireFox 40'
 */
function formatTargetNames(target: Target): string {
  const name = STANDARD_TARGET_NAME_MAPPING[target.target] || target.target;
  return `${name} ${target.version}`;
}

/**
 * Check if a browser version is in the range format
 * ex. 10.0-10.2
 */
function versionIsRange(version: string): boolean {
  return version.includes("-");
}

/**
 * Parse version from caniuse and compare with parsed version from browserslist.
 */
function areVersionsEqual(
  targetVersion: number,
  statsVersion: string
): boolean {
  return targetVersion === parseFloat(statsVersion);
}

/*
 * Check the CanIUse database to see if targets are supported
 *
 * If no record could be found, return true. Rules might not
 * be found because they could belong to another provider
 */
function isSupportedByCanIUse(
  node: AstMetadataApiWithTargetsResolver,
  { version, target, parsedVersion }: Target
): boolean {
  if (!node.caniuseId) return false;

  const data = lite.feature(lite.features[node.caniuseId]);

  if (!data) return true;
  const { stats } = data;
  if (!(target in stats)) return true;

  const targetStats = stats[target];

  if (typeof version === "string" && versionIsRange(version)) {
    return Object.keys(targetStats).some((statsVersion: string): boolean =>
      versionIsRange(statsVersion) &&
      areVersionsEqual(parsedVersion, statsVersion)
        ? !targetStats[statsVersion].includes("y")
        : true
    );
  }

  // @TODO: This assumes that all versions are included in the cainuse db. If this is incorrect,
  //        this will return false negatives. To properly do this, we have to to range comparisons.
  //        Ex. given query for 50 and only version 40 exists in db records, return true
  if (!(version in targetStats)) return true;
  if (!targetStats[version]) return true;

  return targetStats[version].includes("y");
}

/**
 * Return an array of all unsupported targets
 */
export function getUnsupportedTargets(
  node: AstMetadataApiWithTargetsResolver,
  targets: Target[]
): string[] {
  return targets
    .filter((target) => !isSupportedByCanIUse(node, target))
    .map(formatTargetNames);
}

const CanIUseProvider: Array<AstMetadataApiWithTargetsResolver> = [
  // new ServiceWorker()
  {
    caniuseId: "serviceworkers",
    astNodeType: AstNodeTypes.NewExpression,
    object: "ServiceWorker",
  },
  {
    caniuseId: "serviceworkers",
    astNodeType: AstNodeTypes.MemberExpression,
    object: "navigator",
    property: "serviceWorker",
  },
  // document.querySelector()
  {
    caniuseId: "queryselector",
    astNodeType: AstNodeTypes.MemberExpression,
    object: "document",
    property: "querySelector",
  },
  // IntersectionObserver
  {
    caniuseId: "intersectionobserver",
    astNodeType: AstNodeTypes.NewExpression,
    object: "IntersectionObserver",
  },
  // ResizeObserver
  {
    caniuseId: "resizeobserver",
    astNodeType: AstNodeTypes.NewExpression,
    object: "ResizeObserver",
  },
  // PaymentRequest
  {
    caniuseId: "payment-request",
    astNodeType: AstNodeTypes.NewExpression,
    object: "PaymentRequest",
  },
  // Promises
  {
    caniuseId: "promises",
    astNodeType: AstNodeTypes.NewExpression,
    object: "Promise",
  },
  {
    caniuseId: "promises",
    astNodeType: AstNodeTypes.MemberExpression,
    object: "Promise",
    property: "resolve",
  },
  {
    caniuseId: "promises",
    astNodeType: AstNodeTypes.MemberExpression,
    object: "Promise",
    property: "all",
  },
  {
    caniuseId: "promises",
    astNodeType: AstNodeTypes.MemberExpression,
    object: "Promise",
    property: "race",
  },
  {
    caniuseId: "promises",
    astNodeType: AstNodeTypes.MemberExpression,
    object: "Promise",
    property: "reject",
  },
  // fetch
  {
    caniuseId: "fetch",
    astNodeType: AstNodeTypes.CallExpression,
    object: "fetch",
  },
  // document.currentScript()
  {
    caniuseId: "document-currentscript",
    astNodeType: AstNodeTypes.MemberExpression,
    object: "document",
    property: "currentScript",
  },
  // URL
  {
    caniuseId: "url",
    astNodeType: AstNodeTypes.NewExpression,
    object: "URL",
  },
  // URLSearchParams
  {
    caniuseId: "urlsearchparams",
    astNodeType: AstNodeTypes.NewExpression,
    object: "URLSearchParams",
  },
  // performance.now()
  {
    caniuseId: "high-resolution-time",
    astNodeType: AstNodeTypes.MemberExpression,
    object: "performance",
    property: "now",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: AstNodeTypes.NewExpression,
    object: "TypedArray",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: AstNodeTypes.NewExpression,
    object: "Int8Array",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: AstNodeTypes.NewExpression,
    object: "Uint8Array",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: AstNodeTypes.NewExpression,
    object: "Uint8ClampedArray",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: AstNodeTypes.NewExpression,
    object: "Int16Array",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: AstNodeTypes.NewExpression,
    object: "Uint16Array",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: AstNodeTypes.NewExpression,
    object: "Int32Array",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: AstNodeTypes.NewExpression,
    object: "Uint32Array",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: AstNodeTypes.NewExpression,
    object: "Float32Array",
  },
  {
    caniuseId: "typedarrays",
    astNodeType: AstNodeTypes.NewExpression,
    object: "Float64Array",
  },
].map((rule) => ({
  ...rule,
  getUnsupportedTargets,
  id: rule.property ? `${rule.object}.${rule.property}` : rule.object,
  protoChainId: rule.property ? `${rule.object}.${rule.property}` : rule.object,
  protoChain: rule.property ? [rule.object, rule.property] : [rule.object],
}));

export default CanIUseProvider;
