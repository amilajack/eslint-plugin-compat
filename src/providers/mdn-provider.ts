import apiMetadata from "ast-metadata-inferer";
import semver from "semver";
import { ApiMetadata } from "ast-metadata-inferer/lib/types";
import { reverseTargetMappings } from "../helpers";
import { STANDARD_TARGET_NAME_MAPPING } from "../constants";
import { AstMetadataApiWithTargetsResolver, Target } from "../types";
import { BrowserName } from "@mdn/browser-compat-data";

const apis = apiMetadata as ApiMetadata[];

// @TODO Import this type from ast-metadata-inferer after migrating this project to TypeScript
const mdnRecords: Map<string, ApiMetadata> = new Map(
  apis.map((e) => [e.protoChainId, e])
);

interface TargetIdMappings {
  chrome: "chrome";
  firefox: "firefox";
  opera: "opera";
  safari: "safari";
  safari_ios: "ios_saf";
  ie: "ie";
  edge_mobile: "ie_mob";
  edge: "edge";
  opera_android: "and_opera";
  chrome_android: "and_chrome";
  firefox_android: "and_firefox";
  webview_android: "and_webview";
  samsunginternet_android: "and_samsung";
  nodejs: "node";
}

/**
 * Map ids of mdn targets to their "common/friendly" name
 */
const targetIdMappings: Readonly<TargetIdMappings> = {
  chrome: "chrome",
  firefox: "firefox",
  opera: "opera",
  safari: "safari",
  safari_ios: "ios_saf",
  ie: "ie",
  edge_mobile: "ie_mob",
  edge: "edge",
  opera_android: "and_opera",
  chrome_android: "and_chrome",
  firefox_android: "and_firefox",
  webview_android: "and_webview",
  samsunginternet_android: "and_samsung",
  nodejs: "node",
};

const reversedTargetMappings = reverseTargetMappings(targetIdMappings);

/**
 * Take a target's id and return it's full name by using `targetNameMappings`
 * ex. {target: and_ff, version: 40} => 'Android FireFox 40'
 */
function formatTargetNames(target: Target): string {
  return `${STANDARD_TARGET_NAME_MAPPING[target.target]} ${target.version}`;
}

/**
 * Convert '9' => '9.0.0'
 */
function customCoerce(version: string): string {
  return version.length === 1 ? [version, 0, 0].join(".") : version;
}

/*
 * Return if MDN supports the API or not
 */
export function isSupportedByMDN(
  node: AstMetadataApiWithTargetsResolver,
  { version, target: mdnTarget }: Target
): boolean {
  // @ts-ignore
  const target: BrowserName = reversedTargetMappings[mdnTarget];
  // If no record could be found, return true. Rules might not
  // be found because they could belong to another provider
  if (!mdnRecords.has(node.protoChainId)) return true;
  const record = mdnRecords.get(node.protoChainId);
  if (!record || !record.compat.support) return true;
  const compatRecord = record.compat.support[target];
  if (!compatRecord) return true;
  if (!Array.isArray(compatRecord) && !("version_added" in compatRecord))
    return true;
  const { version_added: versionAdded } = Array.isArray(compatRecord)
    ? compatRecord.find((e) => "version_added" in e)!
    : compatRecord;

  // If a version is true then it is supported but version is unsure
  if (typeof versionAdded === "boolean") return versionAdded;
  if (versionAdded === null) return true;

  // Special case for Safari TP: TP is always gte than any other releases
  if (target === "safari") {
    if (version === "TP") return true;
    if (versionAdded === "TP") return false;
  }
  // A browser supports an API if its version is greater than or equal
  // to the first version of the browser that API was added in
  const semverCurrent = semver.coerce(customCoerce(String(version)));
  const semverAdded = semver.coerce(customCoerce(versionAdded));

  // semver.coerce() might be null for non-semvers (other than Safari TP)
  // Just warn and treat features as supported here for now to avoid lint from
  // crashing
  if (!semverCurrent) {
    // eslint-disable-next-line no-console
    console.warn(
      `eslint-plugin-compat: A non-semver target "${target} ${version}" matched for the feature ${node.protoChainId}, skipping. You're welcome to submit this log to https://github.com/amilajack/eslint-plugin-compat/issues for analysis.`
    );
    return true;
  }
  if (!versionAdded) {
    // eslint-disable-next-line no-console
    console.warn(
      `eslint-plugin-compat: The feature ${node.protoChainId} is supported since a non-semver target "${target} ${versionAdded}", skipping. You're welcome to submit this log to https://github.com/amilajack/eslint-plugin-compat/issues for analysis.`
    );
    return true;
  }

  if (!semverAdded) return false;

  return semver.gte(semverCurrent, semverAdded);
}

/**
 * Return an array of all unsupported targets
 */
export function getUnsupportedTargets(
  node: AstMetadataApiWithTargetsResolver,
  targets: Target[]
): string[] {
  return targets
    .filter((target) => !isSupportedByMDN(node, target))
    .map(formatTargetNames);
}

function getMetadataName(metadata: ApiMetadata) {
  switch (metadata.protoChain.length) {
    case 1: {
      return metadata.protoChain[0];
    }
    default:
      return `${metadata.protoChain.join(".")}()`;
  }
}

const MdnProvider: Array<AstMetadataApiWithTargetsResolver> = apis
  // Create entries for each ast node type
  .map((metadata) =>
    metadata.astNodeTypes.map((astNodeType) => ({
      ...metadata,
      name: getMetadataName(metadata),
      id: metadata.protoChainId,
      protoChainId: metadata.protoChainId,
      astNodeType,
      object: metadata.protoChain[0],
      // @TODO Handle cases where 'prototype' is in protoChain
      property: metadata.protoChain[1],
    }))
  )
  // Flatten the array of arrays
  .flat()
  // Add rule and target support logic for each entry
  .map((rule) => ({
    ...rule,
    getUnsupportedTargets,
  }));

export default MdnProvider;
