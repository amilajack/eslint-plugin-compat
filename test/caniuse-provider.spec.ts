import {
  determineTargetsFromConfig,
  parseBrowsersListVersion,
} from "../src/helpers";
import { getUnsupportedTargets } from "../src/providers/caniuse-provider";
import { AstMetadataApiWithTargetsResolver } from "../src/types";
import expectRangeResultJSON from "./expect-range-result-config.json";

describe("CanIUseProvider", () => {
  it("should return unsupported iOS targets with range value for Fetch API", () => {
    const node = { caniuseId: "fetch" } as AstMetadataApiWithTargetsResolver;
    const config = determineTargetsFromConfig(
      ".",
      expectRangeResultJSON.browsers
    );
    const targets = parseBrowsersListVersion(config);
    const result = getUnsupportedTargets(node, targets);
    expect(result).toMatchSnapshot();
  });
});
