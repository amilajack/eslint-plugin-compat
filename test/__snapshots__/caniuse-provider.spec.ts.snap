// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanIUseProvider should return unsupported iOS targets with range value for Fetch API 1`] = `[]`;
